src/test/fuzz/txdownloadman.cpp

```diff
@@ -0,0 +1,408 @@
+    FastRandomContext det_rand{true};
+    node::TxDownloadManager txdownloadman{node::TxDownloadOptions{pool, det_rand, max_orphan_count}};
+
+    std::chrono::microseconds time{244466666};
+
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 10000)
+    {
+        NodeId rand_peer = fuzzed_data_provider.ConsumeIntegralInRange<int64_t>(0, NUM_PEERS);
+
+        // Transaction can be one of the premade ones or a randomly generated one
+        auto rand_tx = fuzzed_data_provider.ConsumeBool() ?
+            MakeTransactionSpending({fuzzed_data_provider.PickValueInArray(COINS)},
+                                    /*num_outputs=*/fuzzed_data_provider.ConsumeIntegralInRange(1, 500),
+                                    /*add_witness=*/fuzzed_data_provider.ConsumeBool()) :
+            TRANSACTIONS.at(fuzzed_data_provider.ConsumeIntegralInRange<unsigned>(0, TRANSACTIONS.size() - 1));
+
+        CallOneOf(
+            fuzzed_data_provider,
+            [&] {
+                node::TxDownloadConnectionInfo info{
+                    .m_preferred = fuzzed_data_provider.ConsumeBool(),
+                    .m_relay_permissions = fuzzed_data_provider.ConsumeBool(),
+                    .m_wtxid_relay = fuzzed_data_provider.ConsumeBool()
+                };
+                txdownloadman.ConnectedPeer(rand_peer, info);
+            },
+            [&] {
+                txdownloadman.DisconnectedPeer(rand_peer);
+                txdownloadman.CheckIsEmpty(rand_peer);
+            },
+            [&] {
+                txdownloadman.ActiveTipChange();
+            },
+            [&] {
+                CBlock block;
+                block.vtx.push_back(rand_tx);
+                txdownloadman.BlockConnected(std::make_shared<CBlock>(block));
+            },
+            [&] {
+                txdownloadman.BlockDisconnected();
+            },
+            [&] {
+                txdownloadman.MempoolAcceptedTx(rand_tx);
+            },
+            [&] {
+                TxValidationState state;
+                state.Invalid(fuzzed_data_provider.PickValueInArray(TESTED_TX_RESULTS), "");
+                txdownloadman.MempoolRejectedTx(rand_tx, state, rand_peer, fuzzed_data_provider.ConsumeBool());
+            },
+            [&] {
+                GenTxid gtxid = fuzzed_data_provider.ConsumeBool() ?
+                                GenTxid::Txid(rand_tx->GetHash()) :
+                                GenTxid::Wtxid(rand_tx->GetWitnessHash());
+                txdownloadman.AddTxAnnouncement(rand_peer, gtxid, time, /*p2p_inv=*/fuzzed_data_provider.ConsumeBool());
+            },
+            [&] {
+                txdownloadman.GetRequestsToSend(rand_peer, time);
+            },
+            [&] {
+                txdownloadman.ReceivedTx(rand_peer, rand_tx);
+                const auto& [should_validate, maybe_package] = txdownloadman.ReceivedTx(rand_peer, rand_tx);
+                Assert(should_validate || !maybe_package.has_value());
```

Comment:
this fails if we have a package to validate...

Reply:
I wrote the assertion incorrectly. The only combination that's not ok is both.